File: qsort.lua
function part (ar,l,r)
    pv = ar[l]
    i = l
    j = r+1
    while true  do
        i=i+1
        while ar[i] <= pv and i <= r do
            i= i+1
        end
        j=j-1
        while ar[j] > pv do
            j= j-1
        end
        if i>=j then break end
        ar[i],ar[j] = ar[j],ar[i]
    end
    ar[l],ar[j] = ar[j],ar[l]
    return j
end    

function quicksort(ar,l,r)
    if l>r then return end
    i = part(ar,l,r);
    quicksort(ar,l,i-1);
    quicksort(ar,i+1,r);
end
   
a={}

math.randomseed(os.time())
for i=1,10000 do
    a[i] = math.random(1,10000)
end

quicksort(a,1,9999);

print(" After sort:");
for i=1,10000 do
    io.write(a[i]);
end

File: qsort.java
import java.util.*;
import java.util.Random;

class qsort{
    static int part(int[] ar, int l, int r){
        int i,j,t;
        int pv = ar[l];
        i = l;
        j = r+1;
        while(true){
            do i++; while (ar[i]<=pv && i<=r);
            do j--; while (ar[j]>pv);
            if (i>=j) break;
            t = ar[i]; ar[i]=ar[j]; ar[j]=t;
        }
        t=ar[l];ar[l]=ar[j];ar[j]=t;
        return j;
        
    }

    static void quicksort(int[] ar,int l, int r){
        int i;
        if (l>r) return;
        i = part(ar,l,r);
        quicksort(ar,l,i-1);
        quicksort(ar,i+1,r);
    }

    public static void  main(String args[]){
        int[] a;
        Random rand = new Random();
        a = new int[10000];    
        for (int i=0;i<10000;i++) 
            a[i] = rand.nextInt(10000);
        System.out.println("");
        quicksort(a,0,9998);
        System.out.println(" After sort:");
        for (int i=0;i<10000;i++) 
            System.out.print(a[i]);
    }
}
